memstore/list/memory_pool.go

```diff
@@ -0,0 +1,171 @@
+func (mp slabMemoryPool) Reallocate(oldBuf [2]uintptr, oldSize int, newSize int) [2]uintptr {
+	if oldSize == newSize {
+		return oldBuf
+	}
+
+	if oldSize != 0 {
+		mp.Free(oldBuf)
+	}
+
+	return mp.Allocate(newSize)
+}
+
+// Free is the implementation of Free in HighLevelMemoryPool interface.
+func (mp slabMemoryPool) Free(buf [2]uintptr) {
+	if !mp.slabAllocator.DecRef(buf) {
+		utils.GetLogger().Panic("buf should be freed but not")
+	}
+}
+
+// Interpret is the implementation of Interpret in HighLevelMemoryPool interface.
+func (mp slabMemoryPool) Interpret(offset uintptr) uintptr {
+	return offset + mp.nativeMemoryPool.GetBaseAddr()
+}
+
+// Destroy is the implementation of Free in HighLevelMemoryPool interface.
+func (mp slabMemoryPool) Destroy() {
+	mp.nativeMemoryPool.Destroy()
+	mp.nativeMemoryPool = nil
+}
+
+// NativeMemoryPool is the interface to manage system memory to support high level memory pool
+// allocation requests. All the pointer/address returned by this memory pool is relative to the
+// base address fetched via GetBaseAddr.
+type NativeMemoryPool interface {
+	// Malloc returns a byte slice to caller, will allocate more memoryOffset if no enough space.
+	// returned addresses are not aligned.
+	Malloc(size int) uintptr
+	// Destroy frees the memory managed by this memory pool. After destroy, any malloc call's
+	// behaviour will be undefined.
+	Destroy()
+	// GetBaseAddr returns the base address managed by this pool.
+	GetBaseAddr() uintptr
+}
+
+// singleChunkNativeMemoryPool manages the system memory as a single chunk of continuous memory address.
+// When this memory pool can no longer satisfy allocation requests, it will allocate another memory chunk
+// big enough to hold the old memory content while satisfying the new requests. Then it will make a memcpy
+// to move the old data into the new location. A small performance penalty will be paid during this period.
+type singleChunkNativeMemoryPool struct {
+	block              uintptr
+	allocatedSize      int
+	totalSize          int
+	nChunks            int
+	hostMemoryReporter HostMemoryReporter
+}
+
+// GetBaseAddr is the implementation of GetBaseAddr in NativeMemoryPool interface.
+func (mp *singleChunkNativeMemoryPool) GetBaseAddr() uintptr {
+	return mp.block
+}
+
+// Malloc is the implementation of Malloc in NativeMemoryPool interface.
+func (mp *singleChunkNativeMemoryPool) Malloc(size int) uintptr {
+	remainingSize := mp.totalSize - mp.allocatedSize
+	if size > remainingSize {
+		newSize := (mp.allocatedSize + size + nativeChunkSize - 1) / nativeChunkSize * nativeChunkSize
+		if mp.hostMemoryReporter != nil {
+			mp.hostMemoryReporter(int64(newSize))
+		}
+		newBlock := uintptr(cgoutils.HostAlloc(newSize))
+		// copy old content.
+		if mp.block != 0 {
+			cgoutils.HostMemCpy(unsafe.Pointer(newBlock), unsafe.Pointer(mp.block), mp.totalSize)
+			cgoutils.HostFree(unsafe.Pointer(mp.block))
+		}
+		mp.block = newBlock
+		mp.totalSize = newSize
+		mp.nChunks += 1
```

Comment:
is this assuming size < 32Mb..?

Reply:
good point,  mp.nChunks should be equal to mp.totalSize / nativeChunkSize